eks_infra/upgrade_cluster.sh

```diff
@@ -0,0 +1,173 @@
+#!/bin/bash
+#/ Usage: 
+#/ export AWS_REGION=<AWS-Region>
+#/ export EC2_KEY_PAIR_NAME=<EC2-Key-Pair-Name>
+#/ export EKS_CLUSTER_MANAGER_ROLE=<ARN-of-IAM-role>
+#/ ./upgrade.sh eks_cluster_name eks_version cluster_autoscalar_image_version
+set -e
+
+# Function to update kubeconfig at ~/.kube/config
+function update_kubeconfig(){
+
+    eksctl utils write-kubeconfig \
+    --cluster ${1} \
+    --authenticator-role-arn ${2} \
+    --region ${3}
+
+    kubectl config get-contexts
+    cat /root/.kube/config
+}
+
+# Function to upgrade eks control plane
+function upgrade_eks_control_plane(){
+
+    eksctl upgrade cluster \
+    --name=${1} \
+    --version ${2} \
+    --approve
+}
+
+# Function to control scaling of cluster autoscalar
+function scale_cluster_autoscalar(){
+    kubectl scale deployments/cluster-autoscaler \
+    --replicas=${1} \
+    -n kube-system
+}
+
+# Function to upgrade autoscalar image
+function upgrade_autoscalar_image(){
+    kubectl -n kube-system \
+    set image deployment.apps/cluster-autoscaler cluster-autoscaler=k8s.gcr.io/autoscaling/cluster-autoscaler:${1}
+}
+
+# Function to create static and dynamic nodegroups in EKS cluster
+function create_nodegroups(){
+    #static nodegroup
+    eksctl create nodegroup \
+    --name static-nodegroup-${2/./-} \
+    --cluster ${1} \
+    --node-type m5.large \
+    --nodes 1 \
+    --node-labels "static=true" \
+    --tags "k8s.io/cluster-autoscaler/node-template/label/static=true" \
+    --asg-access \
+    --ssh-access \
+    --ssh-public-key "${3}"
+
+    #dynamic gpu nodegroup
+    eksctl create nodegroup \
+    --name gpu-nodegroup-${2/./-} \
+    --cluster ${1} \
+    --node-type p3.16xlarge \
+    --nodes-min 0 \
+    --nodes-max 100 \
+    --node-volume-size 80 \
+    --node-labels "test_type=gpu" \
+    --tags "k8s.io/cluster-autoscaler/node-template/label/test_type=gpu" \
+    --asg-access \
+    --ssh-access \
+    --ssh-public-key "${3}"
+
+    # dynamic inf nodegroup
+  
+    eksctl create nodegroup \
+    --name inf-nodegroup-${2/./-} \
+    --cluster ${1} \
+    --node-type inf1.xlarge \
+    --nodes-min 0 \
+    --nodes-max 100 \
+    --node-volume-size 500 \
+    --node-ami "ami-092059396c7e51f52" \
```

Comment:
Please define this node ami as a variable at the top of this script.

Reply:
done